### PR TITLE
feat: harden encoding/decoding functions against panics

### PR DESCRIPTION
**Pull request by Stebalien**
Part of https://github.com/libp2p/go-libp2p/issues/1389

These kinds of functions:

1. Handle user input.
2. Often have out-of-bounds, null pointer, etc bugs.
3. Have completely isolated logic where local panics are unlikely to cause memory corruption elsewhere.

### CODE REVIEW

**Review by marten-seemann — COMMENTED** [review]
As discussed:
* let's keep this to crypto operations
* add stderr output

**Comment by marten-seemann** [comment]
@Stebalien Can you update this PR to log the stack trace to stderr?

File: internal/catch/catch.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package catch
+
+import (
+	"fmt"
+	"os"
+	"runtime/debug"
+)
+
+func CatchAndLog(err *error, where string) {
```
Comment (marten-seemann) [inline internal/catch/catch.go:9]: I like the idea, but I don't think `recover` can be moved to a separate function. I added a random `panic` to one of the functions and it didn't work.

**Review by marten-seemann — CHANGES_REQUESTED** [review]

File: internal/catch/catch.go (hunk 1 above)
Comment (Stebalien) [reply internal/catch/catch.go:9]: This works:

```go
package main

import (
	"fmt"
)

func myRecover() {
	if err := recover(); err != nil {
		fmt.Println("reccovered: ", err)
	}
}

func foo() {
	defer myRecover()

	panic("foo")
}

func main() {
	foo()
}
```


I'll see if I can figure out what I did wrong...

Comment (Stebalien) [reply internal/catch/catch.go:12]: I've added a test case.

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**